tools/clang/unittests/HLSL/ExecutionTest.cpp

```diff
@@ -5770,6 +5799,182 @@ TEST_F(ExecutionTest, WaveIntrinsicsPrefixUintTest) {
       /*isPrefix*/ true);
 }
 
+template <typename T>
+static T GetWaveMultiPrefixInitialAccumValue(LPCWSTR testName) {
+  if (_wcsicmp(testName, L"WaveMultiPrefixProduct") == 0) {
+    return static_cast<T>(1);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixUProduct") == 0) {
+    return static_cast<T>(1);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixSum") == 0) {
+    return static_cast<T>(0);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixUSum") == 0) {
+    return static_cast<T>(0);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitAnd") == 0) {
+    return static_cast<T>(-1);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitOr") == 0) {
+    return static_cast<T>(0);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitXor") == 0) {
+    return static_cast<T>(0);
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixCountBits") == 0) {
+    return static_cast<T>(0);
+  } else {
+    return static_cast<T>(0);
+  }
+}
+
+template <typename T>
+std::function<T(T, T)> GetWaveMultiPrefixReferenceFunction(LPCWSTR testName) {
+  if (_wcsicmp(testName, L"WaveMultiPrefixProduct") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs * rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixUProduct") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs * rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixSum") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs + rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixUSum") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs + rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitAnd") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs & rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitOr") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs | rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixBitXor") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs ^ rhs; };
+  } else if (_wcsicmp(testName, L"WaveMultiPrefixCountBits") == 0) {
+    return [] (T lhs, T rhs) -> T { return lhs + (rhs > 10 ? 1 : 0); };
```

Comment:
What is this 10? I don't understand how that counts bits.

Reply:
It's a bit of a hack to be able to share the same testing infrastructure as the rest of the MultiPrefix tests. The shader code uses `(data.value > 10)` as the boolean input to `WaveMultiPrefxCountBits`. This reference code does the same by adding 1 to the bit count if the value is greater than 10.

Reply:
I see `res.x += WaveMultiPrefixCountBits((input.x == 1), mask);`...? In any case, a comment would help :)

Reply:
Yup, I'll add some comments, and try to clean up this conditional (lots of duplicate bodies).

The shader for this test is in `ShaderOpArithTable.xml`. Specifically: `data.result = WaveMultiPrefixCountBits(data.value > 10, mask);`

The `(input.x == 1)` line is from a compile-only sanity test in `sm_6_5_wave.hlsl`.